src/Pos/Ssc/Extra/Logic.hs

```diff
@@ -181,34 +183,16 @@ type SscGlobalVerifyMode ssc m =
     (MonadSscMem ssc m, SscGStateClass ssc, WithLogger m,
      MonadDB m, MonadError (SscVerifyError ssc) m, WithNodeContext ssc m)
 
-sscRunGlobalUpdatePure
-    :: forall ssc a.
-       LoggerName
-    -> SscGlobalState ssc
-    -> LoggerNameBox (PureLogger (State (SscGlobalState ssc))) a
-    -> (a, SscGlobalState ssc, [LogEvent])
-sscRunGlobalUpdatePure loggerName st =
-    convertRes . flip runState st . runPureLog . usingLoggerName loggerName
-  where
-    convertRes :: ((a, [LogEvent]), SscGlobalState ssc)
-               -> (a, SscGlobalState ssc, [LogEvent])
-    convertRes ((res, events), newSt) = (res, newSt, events)
-
 sscRunGlobalUpdate
     :: forall ssc m a.
        SscGlobalApplyMode ssc m
-    => LoggerNameBox (PureLogger (State (SscGlobalState ssc))) a -> m a
+    => NamedPureLogger (State (SscGlobalState ssc)) a -> m a
 sscRunGlobalUpdate action = do
-    loggerName <- getLoggerName
     globalVar <- sscGlobal <$> askSscMem
-    (res, events) <- atomically $ sscRunGlobalUpdateDo loggerName globalVar
-    res <$ dispatchEvents events
+    launchNamedPureLog (atomically . evalStateTUninitialized) $
+        syncingStateWith globalVar $ switchMonadBaseIdentityToSTM action
   where
-    sscRunGlobalUpdateDo loggerName globalVar = do
-        oldState <- readTVar globalVar
-        let (res, !newState, events) =
-                sscRunGlobalUpdatePure @ssc loggerName oldState action
-        (res, events) <$ writeTVar globalVar newState
+    switchMonadBaseIdentityToSTM = hoist $ hoist $ return . runIdentity
```

Comment:
`return . runIdentity` is [`generalize`](http://hackage.haskell.org/package/mmorph-1.0.9/docs/Control-Monad-Morph.html#v:generalize) from `mmorph`